projects/plugins/jetpack/modules/google-fonts/current/load-google-fonts.php

```diff
@@ -148,6 +150,106 @@ function ( $google_fonts_family ) use ( $available_google_fonts_map, $theme_font
 
 add_filter( 'wp_theme_json_data_default', 'jetpack_register_google_fonts_to_theme_json' );
 
+/**
+ * Filter out the deprecated font families that are from the jetpack-google-fonts provider.
+ *
+ * @param object[] $font_families The font families.
+ * @return object[] The filtered font families.
+ */
+function jetpack_google_fonts_filter_out_deprecated_font_data( $font_families ) {
+	return array_values(
+		array_filter(
+			$font_families,
+			function ( $font_family ) {
+				$has_deprecated_google_fonts_data = false;
+				foreach ( $font_family['fontFace'] as $font_face ) {
+					$provider = isset( $font_face['provider'] ) ? $font_face['provider'] : '';
+					if ( $provider === 'jetpack-google-fonts' ) {
+						$has_deprecated_google_fonts_data = true;
+						break;
+					}
+				}
+
+				return ! $has_deprecated_google_fonts_data;
+			}
+		)
+	);
+}
+
+/**
+ * Unregister the google fonts data from user's theme json data that were stored by accident.
+ *
+ * @param WP_Theme_JSON_Data $theme_json The theme json data of user.
+ * @return WP_Theme_JSON_Data The filtered theme json data.
+ */
+function jetpack_unregister_deprecated_google_fonts_from_theme_json_data_user( $theme_json ) {
+	$raw_data = $theme_json->get_data();
+	$origin   = 'custom';
+	if ( empty( $raw_data['settings']['typography']['fontFamilies'][ $origin ] ) ) {
+		return $theme_json;
+	}
+
+	// Filter out the font definitions that are from the jetpack-google-fonts provider.
+	$raw_data['settings']['typography']['fontFamilies'][ $origin ] = jetpack_google_fonts_filter_out_deprecated_font_data(
+		$raw_data['settings']['typography']['fontFamilies'][ $origin ]
+	);
+
+	$theme_json_class = get_class( $theme_json );
+	return new $theme_json_class( $raw_data, 'custom' );
+}
+
+add_filter( 'wp_theme_json_data_user', 'jetpack_unregister_deprecated_google_fonts_from_theme_json_data_user' );
+
+/**
+ * Filter out the font families from the jetpack-google-fonts provider when saving the user global styles.
+ *
+ * @param array $post_data The post data to filter.
+ * @return array The filtered post data.
+ */
+function jetpack_google_fonts_user_global_styles_pre_save( $post_data ) {
```

Comment:
Do we need this, though? This will always decode and re-encode the saved JSON, for all users, regardless of whether the site actually has bad data.

I think the "more correct" way is to do something like mass-cleanup of all sites with bad user data. Because with this, a site will always have the bad data if it never updates the global styles again.

Since we're already filtering out fonts with deprecated provider, I think we can just leave the bad data as-is. What do you think?

Side note: I'm surprised that the `wp_theme_json_data_user` hook doesn't already filter the `post_data`. It's like we're doing the same logic twice here 😄

Reply:
Yes, it's not necessary. But I'm unsure how to do a mass-cleanup of all atomic sites with bad user data or how to provide a tool for HEs to run when a user encounters this issue...

> I'm surprised that the wp_theme_json_data_user hook doesn't already filter the post_data. It's like we're doing the same logic twice here 😄

I'm surprised too! The endpoint to get/set the user global styles loads the `wp-global-styles` post directly. Whereas, the function to get the global styles relies on the `WP_Theme_JSON_Resolver_Gutenberg` class...

Reply:
We decided to extract the filter-on-save logic to another PR.